lxd/iptables/xtables.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	// outbound <-> container
+	err := ContainerPrepend(protocol, iptablesComment, "nat", "PREROUTING", "-p", connType, "--destination", address, "--dport", port, "-j", "DNAT", "--to-destination", toDest)
+	if err != nil {
+		return err
+	}
+
+	// host <-> container
+	err = ContainerPrepend(protocol, iptablesComment, "nat", "OUTPUT", "-p", connType, "--destination", address, "--dport", port, "-j", "DNAT", "--to-destination", toDest)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NIC Bridged Functions
+
+// NicBridgedRemoveFilters removes any non-standard rules from the nic instance.
+func (xt XTables) NicBridgedRemoveFilters(m deviceConfig.Device, ipv4 net.IP, ipv6 net.IP) error {
+	// Get a current list of rules active on the host.
+	out, err := shared.RunCommand("ebtables", "--concurrent", "-L", "--Lmac2", "--Lx")
+	if err != nil {
+		return fmt.Errorf("Failed to remove network filters for %s: %v", m["name"], err)
+	}
+
+	// Get a list of rules that we would have applied on instance start.
+	rules := generateFilterEbtablesRules(m, ipv4, ipv6)
+
+	errs := []error{}
+	// Iterate through each active rule on the host and try and match it to one the LXD rules.
+	for _, line := range strings.Split(out, "\n") {
+		line = strings.TrimSpace(line)
+		fields := strings.Fields(line)
+		fieldsLen := len(fields)
+
+		for _, rule := range rules {
+			// Rule doesn't match if the field lenths aren't the same, move on.
+			if len(rule) != fieldsLen {
+				continue
+			}
+
+			// Check whether active rule matches one of our rules to delete.
+			if !matchEbtablesRule(fields, rule, true) {
+				continue
+			}
+
+			// If we get this far, then the current host rule matches one of our LXD
+			// rules, so we should run the modified command to delete it.
+			_, err = shared.RunCommand(fields[0], append([]string{"--concurrent"}, fields[1:]...)...)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("Failed to remove network filters rule for %s: %v", m["name"], errs)
+	}
+
+	return nil
+}
+
+// NicBridgedSetFilters sets the nic rules to standard filtering.
+func (xt XTables) NicBridgedSetFilters(m deviceConfig.Device, comment string, ipv4 net.IP, ipv6 net.IP) error {
+	rules := generateFilterEbtablesRules(m, ipv4, ipv6)
+	for _, rule := range rules {
+		_, err := shared.RunCommand(rule[0], append([]string{"--concurrent"}, rule[1:]...)...)
+		if err != nil {
+			return err
+		}
+	}
+
+	rules, err := generateFilterIptablesRules(m, ipv6)
+	if err != nil {
+		return err
+	}
+
+	for _, rule := range rules {
+		err = ContainerPrepend(rule[0], fmt.Sprintf("%s - %s_filtering", comment, rule[0]), "filter", rule[1], rule[2:]...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Network Functions
+
+// NetworkSetupAllowForwarding allows forwarding dependent on boolean argument
+func (xt XTables) NetworkSetupAllowForwarding(protocol string, name string, shouldAccept bool) error {
```

Comment:
Rename protocol to family

Reply:
`shouldAccept` may be better as `action` taking a const (`Action`) defining whether to accept, reject or drop.

Something like:
```
Type Action int
const (
    ActionAccept = Action(1)
    ActionReject = Action(2)
    ActionDrop = Action(3)
)
```

Reply:
If we define the type in firewall then iptables doesn't have access to it in it's xtables implementation. If we define it in iptables then it would be iptables.Action which we could do, but it would be the same type for nftables. Is there an external place we can put the definition of the type?

Reply:
Ah right, good old circular dependencies, aren't those fun...

You could create a `firewall/const` package that would then hold those and may get extended to also include the table names, append/prepend, ... or whatever else we find would be appropriate.

Reply:
Are these constants something you would want extended to the existing iptables package and implementation? Or could we leave those as strings?

Reply:
For now, I'd just do the actions as described above, putting them in a new `firewall/const` package and using them in this branch. We can look at the rest later.

Reply:
I've made this change in the most recent push. I used a slightly different syntax for the const definitions. Let me know if it looks good :)